backend/src/main/java/org/eclipse/tractusx/semantics/registry/service/ShellService.java

```diff
@@ -65,121 +74,139 @@ public Shell save(Shell shell) {
                 .currentPage(pageable.getPageNumber())
-                .totalItems((int)shellsPage.getTotalElements())
+                .totalItems((int) shellsPage.getTotalElements())
                 .totalPages(shellsPage.getTotalPages())
                 .itemCount(shellsPage.getNumberOfElements())
                 .items(shellsPage.getContent())
                 .build();
     }
 
+    private Page<Shell> filterSpecificAssetIdsByTenantId(Page<Shell> shells){
+        String tenantId = tenantAware.getTenantId();
+        return shells.map(shell ->  shell.withIdentifiers(filterSpecificAssetIdsByTenantId(shell.getIdentifiers(), tenantId)));
+    }
+
+    private Set<ShellIdentifier> filterSpecificAssetIdsByTenantId(Set<ShellIdentifier> shellIdentifiers, String tenantId) {
+        // the owning tenant should always see all identifiers
+        if(tenantId.equals(owningTenantId)){
+            return shellIdentifiers;
+        }
+        return shellIdentifiers.stream()
+                .filter(shellIdentifier -> shellIdentifier.getExternalSubjectId() == null ||
+                        shellIdentifier.getExternalSubjectId().equals(tenantId)).collect(Collectors.toSet());
+    }
+
     @Transactional(readOnly = true)
-    public List<String> findExternalShellIdsByIdentifiersByExactMatch(Set<ShellIdentifier> shellIdentifiers){
+    public List<String> findExternalShellIdsByIdentifiersByExactMatch(Set<ShellIdentifier> shellIdentifiers) {
         List<String[]> keyValueCombinations = shellIdentifiers.stream().map(shellIdentifier -> new String[]{shellIdentifier.getKey(), shellIdentifier.getValue()}).collect(Collectors.toList());
-        return shellRepository.findExternalShellIdsByIdentifiersByExactMatch(keyValueCombinations, keyValueCombinations.size());
+        return shellRepository.findExternalShellIdsByIdentifiersByExactMatch(keyValueCombinations,
+                keyValueCombinations.size(), tenantAware.getTenantId(), owningTenantId);
     }
 
     @Transactional(readOnly = true)
-    public List<String> findExternalShellIdsByIdentifiersByAnyMatch(Set<ShellIdentifier> shellIdentifiers){
+    public List<String> findExternalShellIdsByIdentifiersByAnyMatch(Set<ShellIdentifier> shellIdentifiers) {
         List<String[]> keyValueCombinations = shellIdentifiers.stream().map(shellIdentifier -> new String[]{shellIdentifier.getKey(), shellIdentifier.getValue()}).collect(Collectors.toList());
-        return shellRepository.findExternalShellIdsByIdentifiersByAnyMatch(keyValueCombinations);
+        return shellRepository.findExternalShellIdsByIdentifiersByAnyMatch(keyValueCombinations, tenantAware.getTenantId(), owningTenantId);
     }
 
     @Transactional(readOnly = true)
-    public List<Shell> findShellsByExternalShellIds(Set<String> externalShellIds){
+    public List<Shell> findShellsByExternalShellIds(Set<String> externalShellIds) {
         return shellRepository.findShellsByIdExternalIsIn(externalShellIds);
```

Comment:
No filtering required here?

Reply:
Good catch, it's is required here as well.